wpilibj/src/main/java/edu/wpi/first/wpilibj/controller/PIDController.java

```diff
@@ -0,0 +1,639 @@
+      m_deltaTolerance = deltaTolerance;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the percent error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance Percent error which is tolerable.
+   */
+  public void setPercentTolerance(double tolerance) {
+    m_thisMutex.lock();
+    try {
+      m_toleranceType = Tolerance.kPercent;
+      m_tolerance = tolerance;
+      m_deltaTolerance = Double.POSITIVE_INFINITY;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the percent error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance      Percent error which is tolerable.
+   * @param deltaTolerance Change in percent error per second which is tolerable.
+   */
+  public void setPercentTolerance(double tolerance, double deltaTolerance) {
+    m_thisMutex.lock();
+    try {
+      m_toleranceType = Tolerance.kPercent;
+      m_tolerance = tolerance;
+      m_deltaTolerance = deltaTolerance;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the difference between the reference and the measurement.
+   *
+   * @return The error.
+   */
+  public double getError() {
+    m_thisMutex.lock();
+    try {
+      return getContinuousError(m_reference - m_measurementSource.getMeasurement());
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the change in error per second.
+   */
+  public double getDeltaError() {
+    double error = getError();
+
+    m_thisMutex.lock();
+    try {
+      return (error - m_prevError) / getPeriod();
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  @Override
+  @SuppressWarnings("LocalVariableName")
+  public double update() {
+    // Storage for function inputs
+    final double Kp;
+    final double Ki;
+    final double Kd;
+    final double feedforward = m_feedforward.getAsDouble();
+    final double measurement = m_measurementSource.getMeasurement();
+    final double minimumOutput;
+    final double maximumOutput;
+
+    // Storage for function input-outputs
+    final double prevError;
+    final double error;
+    double totalError;
+
+    m_thisMutex.lock();
+    try {
+      Kp = m_Kp;
+      Ki = m_Ki;
+      Kd = m_Kd;
+      minimumOutput = m_minimumOutput;
+      maximumOutput = m_maximumOutput;
+
+      prevError = m_prevError;
+      error = getContinuousError(m_reference - measurement);
+      totalError = m_totalError;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    if (Ki != 0) {
+      totalError = clamp(totalError + error * getPeriod(), minimumOutput / Ki,
+                         maximumOutput / Ki);
+    }
+
+    double output =
+        clamp(Kp * error + Ki * totalError + Kd * (error - prevError) / getPeriod() + feedforward,
+              minimumOutput, maximumOutput);
+
+    m_thisMutex.lock();
+    try {
+      m_prevError = m_error;
+      m_error = error;
```

Comment:
This is the only place `m_error` is assigned. Wouldn't you want to set `m_prevError` to the current error here? Otherwise `m_prevError` would be the error from two steps ago.